packages/skills/weather_client_ledger/handlers.py

```diff
@@ -0,0 +1,230 @@
+                                        sender=self.context.agent_public_key,
+                                        protocol_id=FIPAMessage.protocol_id,
+                                        message=FIPASerializer().encode(msg))
+
+
+class OEFHandler(Handler):
+    """This class scaffolds a handler."""
+
+    SUPPORTED_PROTOCOL = OEFMessage.protocol_id  # type: Optional[ProtocolId]
+
+    def __init__(self, **kwargs):
+        """Initialise the oef handler."""
+        super().__init__(**kwargs)
+        self.dialogue_id = 1
+
+    def setup(self) -> None:
+        """Call to setup the handler."""
+        pass
+
+    def handle(self, message: Message, sender: str) -> None:
+        """
+        Implement the reaction to a message.
+
+        :param message: the message
+        :param sender: the sender
+        :return: None
+        """
+        msg_type = OEFMessage.Type(message.get("type"))
+
+        if msg_type is OEFMessage.Type.SEARCH_RESULT:
+            agents = cast(List[str], message.get("agents"))
+            logger.info("[{}]: found agents={}".format(self.context.agent_name, agents))
+            for agent in agents:
+                msg = FIPAMessage(message_id=STARTING_MESSAGE_ID,
+                                  dialogue_id=self.dialogue_id,
+                                  performative=FIPAMessage.Performative.CFP,
+                                  target=STARTING_TARGET_ID,
+                                  query=None
+                                  )
+                self.dialogue_id += 1
+                self.context.outbox.put_message(to=agent,
+                                                sender=self.context.agent_public_key,
+                                                protocol_id=FIPAMessage.protocol_id,
+                                                message=FIPASerializer().encode(msg))
+
+    def teardown(self) -> None:
+        """
+        Implement the handler teardown.
+
+        :return: None
+        """
+        pass
+
+
+class DefaultHandler(Handler):
+    """This class scaffolds a handler."""
+
+    SUPPORTED_PROTOCOL = DefaultMessage.protocol_id  # type: Optional[ProtocolId]
+
+    def setup(self) -> None:
+        """Call to setup the handler."""
+        pass
+
+    def handle(self, message: Message, sender: str) -> None:
+        """
+        Implement the reaction to a message.
+
+        :param message: the message
+        :param sender: the sender
+        :return: None
+        """
+        json_data = json.loads(message.get("content").decode())
+        if json_data is not None:
+            if "Command" in json_data.keys():
+                if json_data['Command'] == 'success':
+                    logger.info("[{}]: receiving data ...".format(self.context.agent_name))
+                    logger.info("[{}]: this is the data I got: {}".format(self.context.agent_name, json_data))
+                elif json_data['Command'] == "address":
```

Comment:
Can you add a check in both weather clients to check that they only send the transaction if they previously received the data?

Reply:
I can do that but I don't agree. If you think an online transaction, you first pay and then you wait to receive the actual product. What will happen if the client receives the data and then disconnect ?

Reply:
Have you had a look at the Dialogue, Dialogues and DialogueLabel classes in fipa_negotiation? It would probably be good to implement something like this. Otherwise, you can short cut the dialogue etc.